extensions/notebook/src/book/bookModel.ts

```diff
@@ -0,0 +1,221 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+import * as yaml from 'js-yaml';
+import * as glob from 'fast-glob';
+import { BookTreeItem, BookTreeItemType } from './bookTreeItem';
+import { maxBookSearchDepth, notebookConfigKey } from '../common/constants';
+import * as path from 'path';
+import * as fs from 'fs-extra';
+import * as nls from 'vscode-nls';
+import { IJupyterBookToc, IJupyterBookSection } from '../contracts/content';
+
+const localize = nls.loadMessageBundle();
+//const exists = promisify(fs.exists);
+
+export class BookModel implements azdata.nb.NavigationProvider {
+	public BookPath: string;
+	private _bookItems: BookTreeItem[] = [];
+	private _allNotebooks = new Map<string, BookTreeItem>();
+	public openAsUntitled: boolean = false;
+	private _extensionContext: vscode.ExtensionContext;
+	private _tableOfContentPaths: string[] = [];
+	readonly providerId: string = 'BookNavigator';
+	// For testing
+	private _errorMessage: string;
+
+	constructor(bookPath: string, openAsUntitled: boolean, extensionContext: vscode.ExtensionContext) {
+		this.BookPath = bookPath;
+		this.openAsUntitled = openAsUntitled;
+		this._extensionContext = extensionContext;
+		extensionContext.subscriptions.push(azdata.nb.registerNavigationProvider(this));
+		//this.getTableOfContentFiles(this.BookPath).then(() => { this._books = this.readBooks();}, (err) => { console.log(err); });
+
+	}
+
+	public async initializeContents(): Promise<void> {
+		await this.getTableOfContentFiles(this.BookPath);
+		this.readBooks();
+		return;
+	}
+
+	public getAllBooks(): Map<string, BookTreeItem> {
+		return this._allNotebooks;
+	}
+
+	async getTableOfContentFiles(workspacePath: string): Promise<void> {
+		try {
+			let notebookConfig = vscode.workspace.getConfiguration(notebookConfigKey);
+			let maxDepth = notebookConfig[maxBookSearchDepth];
+			// Use default value if user enters an invalid value
+			if (maxDepth === undefined || maxDepth < 0) {
+				maxDepth = 5;
+			} else if (maxDepth === 0) { // No limit of search depth if user enters 0
+				maxDepth = undefined;
+			}
+
+			let p = path.join(workspacePath, '**', '_data', 'toc.yml').replace(/\\/g, '/');
+			let tableOfContentPaths = await glob(p, { deep: maxDepth });
+			this._tableOfContentPaths = this._tableOfContentPaths.concat(tableOfContentPaths);
+			let bookOpened: boolean = this._tableOfContentPaths.length > 0;
+			vscode.commands.executeCommand('setContext', 'bookOpened', bookOpened);
+		} catch (error) {
+			console.log(error);
+		}
+	}
+
+	public readBooks(): BookTreeItem[] {
+		for (const contentPath of this._tableOfContentPaths) {
```

Comment:
is this._tableOfContentPaths is undefined, is an exception thrown here? do we need to error check?

Reply:
Since _tableOfContentPaths is initialized as an empty array, there is never a chance of it being undefined. It's either an empty array or array of paths.